### ISSUE TITLE
Auto-email: "Warning! The Target Kick field is not in the correct format."

### ISSUE THREAD

**Issue by masarapmabuhay** [usbong/UsbongStore#20]
Hi Zent,
I am now creating a new batch to send via auto-email.

For some reason, I am getting the following error:
> Warning! The Target Kick field is not in the correct format.

The target kick field though is in the correct format:
`11/03/2017 12:00 PM
`

What could be cause?

Thank you.

Peace,
Mike

**Comment by zent-lim** [comment]
Afternoon Mike!

Kindly follow the format that is auto generated when the form is loaded:
2017-11-03 12:06:28

Kindly see attached file for screen shot.

Thanks lots!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

The format of Target Kick is as follows:
mm/dd/yyyy hh:mm AM

I am unable to change "AM/PM" to seconds.

The default format must be wrong.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Could you tell me what browser you are using?

Maybe it's a browser form issue.

I'll try to recreate the problem using your browser.

Thanks!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

I am using Google Chrome.
Version 61.0.3163.100 (Official Build) (64-bit)

It did work on Firefox when I tested it in Ubuntu on VirtualBox, so you may be correct.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Thank you for the information.

I was able to recreate the problem.

Please wait a while as I analyze a fix.

Thanks!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

Yes, I was able to use the Create New Batch action button using Firefox.

Peace,
Mike

**Comment by zent-lim** [comment]
Evening Mike,

I sent a pull request to resolve the problem on chrome:
https://github.com/usbong/UsbongStore/pull/22

Please see attached screen shot. You will now be able to save data.

**Comment by masarapmabuhay** [comment]
Hi Zent,
Yes, I have merged your pull request.

Thank you!

Peace,
Mike

**Comment by masarapmabuhay** [comment]
Hi Zent,
I've updated the auto-email code on prod.

Unfortunately, I am now getting the following error message.

> Error Number: 1452
> 
> Cannot add or update a child row: a foreign key constraint fails (`usbong_store`.`auto_email_schedule`, CONSTRAINT `auto_email_schedule_fk_end_customer_id` FOREIGN KEY (`end_customer_id`) REFERENCES `customer` (`customer_id`) ON UPDATE CASCADE)
> 
> INSERT INTO `auto_email_schedule` (`auto_email_id`, `start_customer_id`, `end_customer_id`, `start_datetime`, `status`) VALUES ('2', '1', '30', '2017-11-04 09:16:34', 'QUEUED')
> 
> Filename: models/auto-email/Auto_Email_Schedule_Model.php
> 
> Line Number: 254

May I ask if `usbong_store_auto_email_v05.sql` is really the updated version?

Thank you!

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Did this error occurred when you tried to import v05 sql?

(1) If yes, I dont think you imported the official v05 sql:
https://github.com/usbong/UsbongStore/blob/master/auto-email/usbong_store_auto_email_v05.sql

(2) V05 sql has this line for auto_email_schedule:
INSERT INTO `usbong_store`.`auto_email_schedule` (`auto_email_schedule_id`,
`auto_email_id`, `start_customer_id`, `end_customer_id`, `start_datetime`, `
status`) VALUES (1, 1, 1, 1, '2017-10-1 00:00:00', 'QUEUED');

This has customer id boundaries start = 1 and end =1.

(3) Your import has the ff line:
INSERT INTO auto_email_schedule (auto_email_id, start_customer_id,
end_customer_id, start_datetime, status) VALUES ('2', '1', '30',
'2017-11-04 09:16:34', 'QUEUED')

This has customer id boundaries start = 1 and end =30.

(4) The error was likely caused by the value inserted in `
auto_email_schedule`.`end_customer_id` = 30. Kindly check if your customer
table has an id 30.

If none, please change it to an existing id.

Have a good weekend ahead!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

I've checked prod's customer table. The customer id ends at 13.

It seems that after modifying Start Customer ID and End Customer ID to 2, I was able to create a new batch and add it to the queue.

However, auto-email doesn't send it.
Status remains 0/1 with no errors.

I'm likely doing something wrong.

Thank you!

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

===================
Let's back track a bit.
===================

May I know if the foreign key error occurred when you imported v05 sql? If
yes, your previous concern is closed.

If no, it likely happened when you used UI to create a new batch.

(1) Then there's a problem. The logic I wrote should have caught the error
and failed safely.

(2) If this is the case, please check the ff in your production environment:

File: UsbongStore/usbong_store/application/config/database.php

Field: db_debug

Recommended value: FALSE

Note:

In your environment it's (ENVIRONMENT !== 'production'). This also
evaluates to FALSE when run in prod.

In the past, you changed this to (ENVIRONMENT == 'production') for local
tests and I warned that you should be careful not to reflect this code
update in production server.

Link to Previous Dicsussion:
https://github.com/usbong/UsbongStore/issues/13

===================
No Kick Issue
===================

You need to tie the auto email script in your production environment to
cron.

Below is a link to our old correspondence on this topic:

(1) Link: https://github.com/usbong/UsbongStore/issues/15

(2) Please see the second message from the top in the link above.

In terminal, please use command "crontab -e" to talk with cron.

Link:
https://www.howtogeek.com/101288/how-to-schedule-tasks-on-linux-an-introduction-to-crontab-files/

Please buzz me up for further concerns.

NOTE:
If this is your first broadcast with live server, best to test it with
yourself as recipient first.

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

Yes, I've already imported v05 SQL when the error occurred.

Can you guide me on how I should use "crontab -e" with auto-email in prod?

Thank you.

Peace,
Mike
p.s. Yes, I am testing it with myself as recipient first.

**Comment by zent-lim** [comment]
Hi Mike,

Can you connect to your server with ssh?

===========
IF YES
===========

I do the ff on my end:

(1) Open unix terminal

(2) Key in command below:

ssh root@mystartups.ph

Note:
(a) replace root with your user name
(b) replace mystartups.ph with your domain
(c) judging from your domain name, your command is likely either of the ff:
     - ssh store@usbong.ph
     - ssh root@usbong.ph

(3) The command from (2) will request your password.

(4) If login passes, you're now in a session with your server.

===========
IF NO
===========

Whatever url your host provider has given you will likely have an interface
for you to set up cron.

Maybe check their UI a bit. I'm not familiar with what you are using.

I'm not sure if you are using "Google Cloud Platform".

Quick Google shows how to set up cron in that environment:
https://cloud.google.com/appengine/docs/standard/php/config/cron

Evening!

P.S.
Not familiar with Google Sites.

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

Yes, I can connect to prod server via SSH.

After connecting to prod server, did you want me to use the following command?
`crontab -e`

May I ask what the next steps are?

Thank you!

Peace,
Mike

**Comment by zent-lim** [comment]
Happy sunday Mike!

Pls read on.


On Nov 5, 2017 9:07 AM, "Masarap Mabuhay" <notifications@github.com> wrote:

Hi Zent,
Thank you for the reply.

Yes, I can connect to prod server via SSH.

Ok!

After connecting to prod server, did you want me to use the following
command?
crontab -e

Yes pls.

May I ask what the next steps are?

After u run the command, a terrminal based ediior will open and you ll be
able to edit content.

For sample content, pls see:

(1) Link: https://github.com/usbong/UsbongStore/issues/15

(2) Please see the second message from the top in the link above.

Have a good day ahead!

Thank you!

Peace,
Mike

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/usbong/UsbongStore/issues/20#issuecomment-341941608>,
or mute the thread

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

OK. I ran `crontab -e` on our prod server.
I chose the terminal-based editor, nano, to edit the content.

May I ask what command I should put?

I can recall that I was able to send an auto-email in my instance of Ubuntu on VirtualBox without having to use `crontab -e` at all.

Thank you for your patience.

Peace,
Mike

**Comment by zent-lim** [comment]
Morning Mike!

I ll buzz back when i get access to my laptop around 1 pm. 😊

**Comment by masarapmabuhay** [comment]
OK, Zent!

Thank you.

Peace,
Mike

**Comment by zent-lim** [comment]
Afternoon Mike,

I updated the wiki page for your info request:
https://github.com/usbong/UsbongStore/wiki/Auto-Email-Feature

Please see the intro section for crontab command.

Good afternoon! 😀

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

If I didn't want to have auto-email send emails every fifteen minutes or so, am I right in my understanding that I would only have to run the following command?
`php index.php auto-email/autoemail run`

May I ask how come I have to do this for prod, and not for my Ubuntu instance in VirtualBox?

Thank you!

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Please read on.

On Mon, Nov 6, 2017 at 1:39 PM, Masarap Mabuhay <notifications@github.com>
wrote:

> Hi Zent,
> Thank you for the reply.
>
> If I didn't want to have auto-email send emails every fifteen minutes or
> so, am I right in my understanding that I would only have to run the
> following command?
> php index.php auto-email/autoemail run
>
(1) That's the more tedious option.

(2) The more convenient option is to simply set up the script with cron in
the prescribed manner.

The script will check if you have something to run in your queue.

If there's nothing to run, it wont consume resources.


> May I ask how come I have to do this for prod, and not for my Ubuntu
> instance in VirtualBox?
>

(1) Only you have access to your dev server (Ubuntu instance in
VirtualBox). As such, customers have no access. If a new customer
registers, your dev environment will not detect the registration.

(2) The most updated customer data and product data are in the production
server and not in your development server. If you use your dev server to
create the email, the ids and links generated might not match the
production server links.


> Thank you!
>
You're welcome!

Please buzz me up for further concerns.

Afternoon!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

1) If I wanted to send my auto-email only once, do I still need to set my cron script?

2) I was able to send auto-emails in my Ubuntu on VirtualBox, without having to use `cron`. Could it be that `cron` is used for sending regular auto-emails with the same content, and not for individual auto-emails with different contents? 

Thank you for the clarifications.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Please read on.

On Mon, Nov 6, 2017 at 3:40 PM, Masarap Mabuhay <notifications@github.com>
wrote:

> Hi Zent,
> Thank you for the reply.
>
>    1.
>
>    If I wanted to send my auto-email only once, do I still need to set my
>    cron script?
>
> Do you mean just one email to one person in Usbong's life time?

If yes, there is no need to set up the script.

>
>    1.
>    2.
>
>    I was able to send auto-emails in my Ubuntu on VirtualBox, without
>    having to use cron. Could it be that cron is used for sending regular
>    auto-emails with the same content, and not for individual auto-emails with
>    different contents?
>
> Nope.

Cron is used to send "individual auto-emails with different contents".

The auto email script is state aware.

Each time you run the script, it does something different depending on the
state of your database.

The script will know what email has already been sent out. It will not send
out the same email to the same person unless you explicitly instruct it to
do so.

Feel free to buzz for further concerns.

Have a good day ahead!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

1) OK. The issue that I'm currently experiencing with auto-email though is that the email is not sent at all. Let me test it again.

2) OK. Thank you for clarifying how I should use `cron`.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

I think we're going over the same problems repeatedly.

Would it help if we just met and I set up your production server?

Have a good day ahead!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

I can make time for us to meet this Thurs (11/9), 2pm.
Please let me know if you're available.
We can meet at the Usbong HQ.

In the meantime, you can create a New Account on prod, and test auto-email.
https://store.usbong.ph/account/create/

I can upgrade your account to admin, so that you'll have access to auto-email.

Please let me know if this would be OK with you.

Thank you.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

Sorry. Thursday is full.

I m checking my sunday p.m.

Waiting for reply from family.

Would a sunday p.m. work for you?

Thanks!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

We can schedule it the following week, Thurs (Nov. 16), 2pm.
However, if it is possible for us to resolve the issue online, that may be OK too.

I understand that your time is precious, and for this case, we may be able to already figure out a fix remotely. 

My hunch is that you'll only need to create a New Account on prod, and test auto-email.
https://store.usbong.ph/account/create/

Please let me know if you're OK with the latter, so that I can give your account admin access.

Thank you.

Peace,
Mike

**Comment by zent-lim** [comment]
Hi Mike,

(1) Other than admin access to front end UI, I'll also be needing ssh
access to your live server (set up cron script and look at log files).

The later is very sensitive information.

Additionally, it needs to be supervised by you as admin access to server
files might be catastrophic if handled incorrectly.

(2) On the other hand, I'm ok resolving issues remotely. Can we continue
looking at the cron script after we settle Issue 23 below?

https://github.com/usbong/UsbongStore/issues/23#issuecomment-342448637

Have a good day ahead!

**Comment by masarapmabuhay** [comment]
Hi Zent,
Thank you for the reply.

1) We may actually need not have to set up cron. I am already OK with sending individual auto-emails with a press of a button.

2) Yes, it seems that your solution (https://github.com/usbong/UsbongStore/issues/23#issuecomment-342448637) has resolved the issue.

In the meantime, I have put this as a new issue here (https://github.com/usbong/UsbongStore/issues/24), so that we can already close this one.

Thank you.

Peace,
Mike

**Comment by zent-lim** [comment]
Okly dokly! 😁

On Nov 8, 2017 8:25 AM, "Masarap Mabuhay" <notifications@github.com> wrote:

Closed #20 <https://github.com/usbong/UsbongStore/issues/20>.

—
You are receiving this because you were mentioned.

Reply to this email directly, view it on GitHub
<https://github.com/usbong/UsbongStore/issues/20#event-1330891027>, or mute
the thread
